brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClientTracing.java

```diff
@@ -0,0 +1,139 @@
+import static com.github.kristofa.brave.internal.Util.checkNotNull;
+
+public final class HttpClientTracing {
+
+  /** Creates a tracing interceptor with defaults. Use {@link #builder(Tracing)} to customize. */
+  public static HttpClientTracing create(Tracing tracing) {
+    return builder(tracing).build();
+  }
+
+  public static Builder builder(Tracing tracing) {
+    return new Builder(tracing);
+  }
+
+  public static final class Builder {
+    final Tracing tracing;
+    HttpClientParser<HttpRequestWrapper, CloseableHttpResponse> parser = new HttpClientParser<>();
+
+    Builder(Tracing tracing) { // intentionally hidden
+      this.tracing = checkNotNull(tracing, "tracing");
+    }
+
+    public Builder parser(HttpClientParser<HttpRequestWrapper, CloseableHttpResponse> parser) {
+      this.parser = checkNotNull(parser, "parser");
+      return this;
+    }
+
+    public HttpClientTracing build() {
+      return new HttpClientTracing(this);
+    }
+  }
+
+  final Tracer tracer;
+  final HttpClientHandler<HttpRequestWrapper, CloseableHttpResponse> handler;
+  final TraceContext.Injector<HttpMessage> injector;
+
+  HttpClientTracing(Builder b) {
+    this.tracer = b.tracing.tracer();
+    this.handler =
+        new HttpClientHandler<HttpRequestWrapper, CloseableHttpResponse>(new HttpAdapter(),
+            b.parser);
+    this.injector = b.tracing.propagation().injector(HttpMessage::setHeader);
+  }
+
+  /**
+   * Preferred way for constructing http clients, as it allows use of {@link Tracer#currentSpan}
+   *
+   * <p>Implicitly requires a minimum version of httpclient 4.3
+   */
+  public HttpClientBuilder httpClientBuilder() {
+    return new HttpClientBuilder() {
+
+      /**
+       * protocol exec is the second in the execution chain, so is invoked before a request is
+       * provisioned. We provision and scope a span here, so that application interceptors can see
+       * it via {@link Tracer#currentSpan()}.
+       */
+      @Override protected ClientExecChain decorateProtocolExec(ClientExecChain exec) {
+        return (route, request, context, execAware) -> {
+          Span next = tracer.nextSpan();
+          context.setAttribute(SpanInScope.class.getName(), tracer.withSpanInScope(next));
+          try {
+            return exec.execute(route, request, context, execAware);
+          } catch (IOException | HttpException | RuntimeException e) {
+            context.getAttribute(SpanInScope.class.getName(), SpanInScope.class).close();
+            throw e;
+          }
+        };
+      }
+
+      /**
+       * main exec is the first in the execution chain, so last to execute. This creates a concrete
+       * http request, so this is where the timing in the span occurs.
+       *
+       * <p>This ends the span (and scoping of it) created by {@link #decorateMainExec(ClientExecChain)}.
+       */
+      @Override protected ClientExecChain decorateMainExec(ClientExecChain exec) {
```

Comment:
@llinder @devinsba @hyleung so when introducing support for custom parsing, I chose to keep generic parameters "raw" meaning they represent the actual native types. This is mostly to make the below code readable and without the clutter of `new FooRequestAdapter(request)` and `new FooResponseAdapter(response)`.

You'll also notice the adapter (maybe converter?) below is very straightforward.